### PR TITLE
Add basic support for Météo France API

### PR DESCRIPTION
**Pull request by papjul**
Météo France is the French national meteorological service.
Here is a basic support for their unofficial API.

What works:
- Search for a location (query or coordinates)
- Daily forecast for the next 15 days
- Hourly forecast
- Humidity
- Clouds
- Sunset and sunrise
- Partial support for wind
- Partial support for rain
- Partial support for official authorities weather warnings
- French translation was improved and completed

What does not work/is not available yet:
- There is no Air Quality data. This is the same as AccuWeather (there is no Air Quality support for France). However, Météo France uses Air Parif data to display air quality information in their official app so we might use this in the future, although it only works for "Île de France" region so it's only a minor addition to look forward to.
- There is a "Feels like null" displayed. This is a bug at app level (we need to add "when feels_like is null, do not display").
- Rain for the next 60 minutes: this is one of the first things that I wish to add for the supported cities.
- Warnings do not work if using current location with GPS instead of manual location, because app is using data from Accu Weather instead of Météo France and Accu Weather does not process regions the same way (warnings are issued "region"-wide). App has a strange way of mixing location searches through all providers. We may be able to counter this by getting the "domain" returned by the daily forecast "position" information and then make an additional API call.
- The "night time" information only contains the temperature, the rest is not available with daily API. We should be able to retrieve it through the hourly weather data but this will only cover about 7 days (hourly forecast does not go beyond), not the whole 15 days. So we need to address this at app level (make night @Nullable and "if HalfDay night is null, do not display night information in daily chart").
- Complete support for wind with the hourly weather data, because it is not available at all in the daily weather data from API. Same way of proceeding as above.
- Moonset and moonrise: this information is only available for the current day with an additional API query, so we may be able to add it in the future. Not really urgent.

To sum up, some minor things are missing from the set of data Météo France provides for free, so we don't have a full set of data in Geometric Weather, unlike what is possible Accu Weather, but Météo France data is still more accurate (in France), so adding this API is a good addition to this wonderful app.

I added the "(beta)" keyword to the weather data provider name due to the few things that need to be addressed, but in the meantime everything in the "What works" section should already be sufficient for daily usage.

Note: I had to update GreenDAO module because I could not compile it otherwise.

### CODE REVIEW

**Comment by mbestavros** [comment]
@papjul This is absolutely awesome! Thanks for the contribution.

I'd encourage you to propose this [upstream](https://github.com/WangDaYeeeeee/GeometricWeather) so people who have the app through the Play Store also have this option. 

If they end up not being responsive, I'd be happy to add this to my fork in some way.

BTW, we should connect about further enhancements -- I'd love to get another "broad-spectrum" API available as an alternative to AccuWeather, like [OpenWeatherMap](https://openweathermap.org/). I gave it a stab, but didn't make much progress. If you'd be willing to give that a try as well, that would be fantastic.

**Comment by papjul** [comment]
Thank you for your support!

I'm not sure about upstream, many things were removed in your version, which would prove difficult to maintain in two different versions (mainly due to provider list being different and being heavily tied to the core of the app, as opposed to providers as extensions (check the way [Tachiyomi extensions](/inorichi/tachiyomi-extensions) work for example)), especially since there are no tests. It would require manual testing in the two versions, notably in one that I do not plan to use at all: people installing the original version are probably more interested by the Chinese providers.

For Play Store users, the question would more likely be: should we maintain this fork as a different "Western" app? So it could be published through Play Store (similar to Vinyl Music Player vs Phonograph).
There has not been a lot of commits recently on the original Geometric Weather so we can still cherry-pick some commits from time to time.

I can probably help on OpenWeatherMap. Implementing the Météo France API happened to be very easy, it requires no Android knowledge, just some basic Java (and some copy/paste/adapt from the already existing API). The most difficult part is the availability of data since Geometric Weather requires a lot of mandatory data which is not always available.

After seeing so many junk weather apps, having one that just does its job without a clunky interface, and open source, feels so good, so I'm willing to spend some time enhancing it. We can connect but I don't know what's the best way for a small team today, I know there are some cool open source projects out there for this (Mattermost, Element, etc) or we can just use some Discord, please let me know your choice.

**Comment by mbestavros** [comment]
@papjul All great questions/points! I'll try setting a project chat up this weekend (that's right up my alley) and we can discuss further.

> After seeing so many junk weather apps, having one that just does its job without a clunky interface, and open source, feels so good, so I'm willing to spend some time enhancing it.

You read my mind! Exactly why I'm so interested in making this a thing.

**Comment by mbestavros** [comment]
@papjul I'm on my holiday break, so I've finally found the time to actually do what I said 😅 

I created a community and room on matrix.org where we can chat: https://matrix.to/#/+geometric-weather-foss:matrix.org

Hope to see you there! We can discuss other contribution details as well. I'm also going to try and figure out a good way to merge and release this on the F-Droid app, since it's really good work.

File: app/src/main/java/wangdaye/com/geometricweather/basic/model/option/utils/OptionMapper.java [hunk 1]
```diff
@@ -114,15 +117,15 @@ public static DarkMode getDarkMode(String value) {
                 return DarkMode.AUTO;
         }
     }
-    
+
```
Comment (mbestavros) [inline app/src/main/java/wangdaye/com/geometricweather/basic/model/option/utils/OptionMapper.java:120]: I'd prefer we not introduce spacing changes in large commits/PRs like this. Could you remove these from your first commit?

File: app/src/main/java/wangdaye/com/geometricweather/basic/model/option/utils/OptionMapper.java [hunk 2]
```diff
@@ -114,15 +117,15 @@ public static DarkMode getDarkMode(String value) {
...
     public static TemperatureUnit getTemperatureUnit(String value) {
         switch (value) {
             case "f":
                 return TemperatureUnit.F;
 
             case "k":
                 return TemperatureUnit.K;
-                
+
```
Comment (mbestavros) [inline app/src/main/java/wangdaye/com/geometricweather/basic/model/option/utils/OptionMapper.java:128]: Remove spacing, please.

File: app/src/main/java/wangdaye/com/geometricweather/basic/model/option/utils/OptionMapper.java [hunk 3]
```diff
@@ -203,12 +206,12 @@ public static SpeedUnit getSpeedUnit(String value) {
                 return SpeedUnit.KPH;
         }
     }
-    
+
```
Comment (mbestavros) [inline app/src/main/java/wangdaye/com/geometricweather/basic/model/option/utils/OptionMapper.java:209]: Remove spacing, please.

File: app/src/main/java/wangdaye/com/geometricweather/basic/model/option/utils/OptionMapper.java [hunk 4]
```diff
@@ -203,12 +206,12 @@ public static SpeedUnit getSpeedUnit(String value) {
...
     public static UIStyle getUIStyle(String value) {
         switch (value) {
             case  "circular":
                 return UIStyle.CIRCULAR;
-                
+
```
Comment (mbestavros) [inline app/src/main/java/wangdaye/com/geometricweather/basic/model/option/utils/OptionMapper.java:214]: Remove spacing, please.

File: app/src/main/java/wangdaye/com/geometricweather/basic/model/option/utils/OptionMapper.java [hunk 5]
```diff
@@ -341,7 +344,7 @@ public static String getDailyTrendDisplaySummary(Context context, @NonNull List<
         }
         return builder.toString().replace(",", ", ");
     }
-    
+
```
Comment (mbestavros) [inline app/src/main/java/wangdaye/com/geometricweather/basic/model/option/utils/OptionMapper.java:347]: Remove spacing, please.

File: app/src/main/java/wangdaye/com/geometricweather/location/LocationHelper.java [hunk 6]
```diff
@@ -186,4 +224,4 @@ public void requestLocationSuccess(String query, List<Location> locationList) {
     public void requestLocationFailed(String query) {
         listener.requestLocationFailed(location);
     }
-}
\ No newline at end of file
+}
```
Comment (mbestavros) [inline app/src/main/java/wangdaye/com/geometricweather/location/LocationHelper.java:227]: I appreciate adding newlines is a good thing, but perhaps in a separate commit.

File: app/src/main/res/values-cs/arrays.xml [hunk 7]
```diff
@@ -184,4 +185,4 @@
         <item>/m³</item>
     </string-array>
 
-</resources>
\ No newline at end of file
+</resources>
```
Comment (mbestavros) [inline app/src/main/res/values-cs/arrays.xml:188]: Same here -- ending newlines are good, but in a separate commit/PR.

File: app/src/main/res/values-ja/arrays.xml [hunk 8]
```diff
@@ -136,4 +137,4 @@
         <item>夜間</item>
     </string-array>
 
-</resources>
\ No newline at end of file
+</resources>
```
Comment (mbestavros) [inline app/src/main/res/values-ja/arrays.xml:140]: Remove this newline, please.

File: app/src/main/res/values-ko/arrays.xml [hunk 9]
```diff
@@ -184,4 +185,4 @@
         <item>/m³</item>
     </string-array>
 
-</resources>
\ No newline at end of file
+</resources>
```
Comment (mbestavros) [inline app/src/main/res/values-ko/arrays.xml:188]: Remove this newline, please.

File: app/src/main/res/values-pl/arrays.xml [hunk 10]
```diff
@@ -183,4 +184,4 @@
         <item>/m³</item>
     </string-array>
 
-</resources>
\ No newline at end of file
+</resources>
```
Comment (mbestavros) [inline app/src/main/res/values-pl/arrays.xml:187]: Remove this newline, please.

File: app/src/main/res/values-ru/arrays.xml [hunk 11]
```diff
@@ -181,4 +182,4 @@
         <item>/м³</item>
     </string-array>
 
-</resources>
\ No newline at end of file
+</resources>
```
Comment (mbestavros) [inline app/src/main/res/values-ru/arrays.xml:185]: Remove this newline, please.

File: app/src/main/res/values-zh-rCN/arrays.xml [hunk 12]
```diff
@@ -181,4 +182,4 @@
         <item>/米³</item>
     </string-array>
 
-</resources>
\ No newline at end of file
+</resources>
```
Comment (mbestavros) [inline app/src/main/res/values-zh-rCN/arrays.xml:185]: Remove this newline, please.

File: app/src/main/res/values-zh-rHK/arrays.xml [hunk 13]
```diff
@@ -181,4 +182,4 @@
         <item>/米³</item>
     </string-array>
 
-</resources>
\ No newline at end of file
+</resources>
```
Comment (mbestavros) [inline app/src/main/res/values-zh-rHK/arrays.xml:185]: Remove this newline, please.

File: app/src/main/res/values-zh-rTW/arrays.xml [hunk 14]
```diff
@@ -181,4 +182,4 @@
         <item>/米³</item>
     </string-array>
 
-</resources>
\ No newline at end of file
+</resources>
```
Comment (mbestavros) [inline app/src/main/res/values-zh-rTW/arrays.xml:185]: Remove this newline, please.

File: app/src/main/res/values/arrays.xml [hunk 15]
```diff
@@ -423,4 +425,4 @@
         <item>ppcm</item>
     </string-array>
 
-</resources>
\ No newline at end of file
+</resources>
```
Comment (mbestavros) [inline app/src/main/res/values/arrays.xml:428]: Remove this newline, please.

File: gradle.properties [hunk 16]
```diff
@@ -19,5 +19,15 @@ ACCU_CURRENT_KEY="131778526309453295c9ce2350a79e87"
...
+IQA_ATMO_AURA_KEY="e139ed41f812383ed88678f8e7fa744f"
+IQA_ATMO_AURA_URL="https\://api.atmo-aura.fr/"
+#IQA_ATMO_SUD_KEY="NO_API_KEY_REQUIRED"
+IQA_ATMO_SUD_URL="https\://api.atmosud.org/iqa/"
+# More IQA regions could be added: https://atmo-france.org/la-qualite-de-lair-dans-votre-region
+
 android.useAndroidX=true
-android.enableJetifier=true
\ No newline at end of file
+android.enableJetifier=true
```
Comment (mbestavros) [inline gradle.properties:33]: Remove this newline, please.

**Review by mbestavros — CHANGES_REQUESTED** [review]
@papjul Thanks so much for the continued cooperation in getting this merge-ready! I've done one final review -- you introduced a lot of spacing changes in this PR that, while I agree are good, would probably be best introduced as a separate patch. As-is, it means that there are lots of files where the only modification is just a newline inserted at the end, which is bad for auditability.

Given that, I have two requests for changes:

1: Could you edit your first commit to remove these spacing changes? I've highlighted them inline.
2: Could you edit the wording of your first commit to better reflect the (tremendous) work done? Right now it just reads as "Start implementing some things". Something like "Implement beta support for Meteo France API" would be perfect.

If that's done, I think this looks good to go, and will happily merge -- very excited 😄

**Review by mbestavros — APPROVED** [review]
I've changed my mind about the spacing changes -- they're too annoying to undo, and I'd like to get this merged in so @papjul can get cracking on other things. The commit wording looks good now, so I think this PR is ready to go!
